can you try this change on top of #5522 and run sanitycheck with

`# sanitycheck -p qemu_x86 -N`

I am getting lots of failures. for example:

```
FAILED: zephyr/zephyr_prebuilt.elf
: && ccache /home/nashif/Work/SDK/0.9.2/sysroots/x86_64-pokysdk-linux/usr/bin/i586-zephyr-elf/i586-zephyr-elf-gcc    zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj  -o zephyr/zephyr_prebuilt.elf  -T zephyr/linker.cmd -Wl,-Map=/home/nashif/Work/zephyr/samples/net/coap_server/build/zephyr/zephyr.map -u_OffsetAbsSyms -u_ConfigAbsSyms -e__start -Wl,--start-group -Wl,--whole-archive libapp.a zephyr/libzephyr.a zephyr/subsys/bluetooth/common/libsubsys__bluetooth__common.a zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a zephyr/subsys/net/libsubsys__net.a zephyr/subsys/net/ip/libsubsys__net__ip.a zephyr/subsys/net/ip/l2/libsubsys__net__ip__l2.a zephyr/arch/x86/soc/ia32/libarch__x86__soc__ia32.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a zephyr/CMakeFiles/offsets.dir/arch/x86/core/offsets/offsets.c.obj -Wl,--end-group -L"/home/nashif/Work/SDK/0.9.2/sysroots/i586-zephyr-elf/usr/lib/i586-zephyr-elf/6.2.0" -L/home/nashif/Work/zephyr/samples/net/coap_server/build/zephyr -lgcc -Wl,--print-memory-usage -nostartfiles -nodefaultlibs -nostdlib -static -no-pie -Wl,-X -Wl,-N -Wl,--gc-sections -Wl,--build-id=none && :
Memory region         Used Size  Region Size  %age Used
             ROM:      138843 B      4092 KB      3.31%
             RAM:        108 KB         4 MB      2.64%
        IDT_LIST:         492 B         2 KB     24.02%
        MMU_LIST:        9296 B         1 KB    907.81/home/nashif/Work/SDK/0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: zephyr/zephyr_prebuilt.elf section `.noinit."/home/nashif/Work/zephyr/subsys/bluetooth/host/hci_core.c".2' will not fit in region `MMU_LIST'
/home/nashif/Work/SDK/0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region `MMU_LIST' overflowed by 8272 bytes
collect2: error: ld returned 1 exit status
%
ninja: build stopped: subcommand failed.
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -467,7 +467,17 @@ if(CONFIG_APPLICATION_MEMORY)
       string(SUBSTRING ${fixed_path} 1 -1 fixed_path)
     endif()
 
-    list(APPEND ks "${fixed_path}lib${target_name}.a")
+    set(fixed_path "${fixed_path}lib${target_name}.a")
+
+    if(CMAKE_GENERATOR STREQUAL "Ninja")
+      # Ninja invokes the linker from the root of the build directory
+      # (APPLICATION_BINARY_DIR) instead of from the build/zephyr
+      # directory (PROJECT_BINARY_DIR). So for linker-defs.h to get
+      # the correct path we need to prefix with zephyr/.
+      set(fixed_path "zephyr/${fixed_path}")
+    endif()
+
+    list(APPEND ks ${fixed_path})
   endforeach()
 
   # We are done constructing kernel_object_file_list, now we inject this

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -500,24 +500,52 @@ if(CONFIG_APPLICATION_MEMORY)
   endforeach()
 endif() # CONFIG_APPLICATION_MEMORY
 
+function(construct_add_custom_command_for_linker_pass linker_pass_number output_variable)
+  if(linker_pass_number EQUAL 1)
+    set(is_first_pass 1)
+  elseif(linker_pass_number GREATER 1)
+    set(is_first_pass 0)
+  else()
+    assert(0 "Unreachable code")
+  endif()
+
+  if(is_first_pass)
+    set(linker_cmd_file_name linker.cmd)
+  else()
+    set(linker_cmd_file_name linker_pass${linker_pass_number}.cmd)
+  endif()
+
+  if(is_first_pass)
+    set(LINKER_PASS_DEFINE "")
+  else()
+    set(LINKER_PASS_DEFINE -DLINKER_PASS${linker_pass_number})
+  endif()
+
+  set(${output_variable}
+    OUTPUT ${linker_cmd_file_name}
+    DEPENDS ${LINKER_SCRIPT}
+    ${LINKER_SCRIPT_DEP}
+    COMMAND ${CMAKE_C_COMPILER}
+    -x assembler-with-cpp
+    ${NOSTDINC_F}
+    -undef
+    -MD -MF ${linker_cmd_file_name}.dep -MT ${BASE_NAME}/${linker_cmd_file_name}
+    ${ZEPHYR_INCLUDES}
+    ${LINKER_SCRIPT_DEFINES}
+    ${LINKER_PASS_DEFINE}
+    -E ${LINKER_SCRIPT} -P
+    -o ${linker_cmd_file_name}
+    VERBATIM
+    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+
+    PARENT_SCOPE
+    )
+endfunction()
+
 get_filename_component(BASE_NAME ${CMAKE_CURRENT_BINARY_DIR} NAME)
+construct_add_custom_command_for_linker_pass(1 custom_command)
 add_custom_command(
-  OUTPUT linker.cmd
-  DEPENDS ${LINKER_SCRIPT}
-  ${LINKER_SCRIPT_DEP}
-  # NB: This COMMAND is copy-pasted to generate linker_pass2.cmd
-  # TODO: Remove duplication
-  COMMAND ${CMAKE_C_COMPILER}
-  -x assembler-with-cpp
-  ${NOSTDINC_F}
-  -undef
-  -MD -MF linker.cmd.dep -MT ${BASE_NAME}/linker.cmd
-  ${ZEPHYR_INCLUDES}
-  ${LINKER_SCRIPT_DEFINES}
-  -E ${LINKER_SCRIPT} -P
-  -o linker.cmd
-  VERBATIM
-  WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+  ${custom_command}
 )
 add_custom_target(
   linker_script
@@ -701,24 +729,11 @@ if(GKOF OR GKSF)
   set(logical_target_for_zephyr_elf kernel_elf)
 
   # The second linker pass uses the same source linker script of the
-  # first pass (LINKER_SCRIPT), but this time preprocessed with the
-  # define LINKER_PASS2.
+  # first pass (LINKER_SCRIPT), but this time with a different output
+  # file and preprocessed with the define LINKER_PASS2.
+  construct_add_custom_command_for_linker_pass(2 custom_command)
   add_custom_command(
-    OUTPUT linker_pass2.cmd
-    DEPENDS ${LINKER_SCRIPT}
-    ${LINKER_SCRIPT_DEP}
-    COMMAND ${CMAKE_C_COMPILER}
-    -x assembler-with-cpp
-    ${NOSTDINC_F}
-    -undef
-    -MD -MF linker_pass2.cmd.dep -MT ${BASE_NAME}/linker_pass2.cmd
-    ${ZEPHYR_INCLUDES}
-    ${LINKER_SCRIPT_DEFINES}
-    -DLINKER_PASS2
-    -E ${LINKER_SCRIPT} -P
-    -o linker_pass2.cmd
-    VERBATIM
-    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+    ${custom_command}
     )
   add_custom_target(
     linker_pass2_script

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -396,25 +396,6 @@ endforeach()
 
 get_property(OUTPUT_FORMAT        GLOBAL PROPERTY PROPERTY_OUTPUT_FORMAT)
 
-# Run the pre-processor on the linker script
-#
-# Deal with the un-preprocessed linker scripts differently with
-# different generators.
-if(CMAKE_GENERATOR STREQUAL "Unix Makefiles")
-  # Note that the IMPLICIT_DEPENDS option is currently supported only
-  # for Makefile generators and will be ignored by other generators.
-  set(LINKER_SCRIPT_DEP IMPLICIT_DEPENDS C ${LINKER_SCRIPT})
-elseif(CMAKE_GENERATOR STREQUAL "Ninja")
-  # Using DEPFILE with other generators than Ninja is an error.
-  set(LINKER_SCRIPT_DEP DEPFILE ${PROJECT_BINARY_DIR}/linker.cmd.dep)
-else()
-  # TODO: How would the linker script dependencies work for non-linker
-  # script generators.
-  message(STATUS "Warning; this generator is not well supported. The
-  Linker script may not be regenerated when it should.")
-  set(LINKER_SCRIPT_DEP "")
-endif()
-
 get_property(LINKER_SCRIPT_DEFINES GLOBAL PROPERTY PROPERTY_LINKER_SCRIPT_DEFINES)
 
 if(CONFIG_APPLICATION_MEMORY)
@@ -519,6 +500,22 @@ function(construct_add_custom_command_for_linker_pass linker_pass_number output_
     set(LINKER_PASS_DEFINE "")
   else()
     set(LINKER_PASS_DEFINE -DLINKER_PASS${linker_pass_number})
+  endif()
+
+  # Different generators deal with depfiles differently.
+  if(CMAKE_GENERATOR STREQUAL "Unix Makefiles")
+    # Note that the IMPLICIT_DEPENDS option is currently supported only
+    # for Makefile generators and will be ignored by other generators.
+    set(LINKER_SCRIPT_DEP IMPLICIT_DEPENDS C ${LINKER_SCRIPT})
+  elseif(CMAKE_GENERATOR STREQUAL "Ninja")
+    # Using DEPFILE with other generators than Ninja is an error.
+    set(LINKER_SCRIPT_DEP DEPFILE ${PROJECT_BINARY_DIR}/${linker_cmd_file_name}.dep)
+  else()
+    # TODO: How would the linker script dependencies work for non-linker
+    # script generators.
+    message(STATUS "Warning; this generator is not well supported. The
+  Linker script may not be regenerated when it should.")
+    set(LINKER_SCRIPT_DEP "")
   endif()
 
   set(${output_variable}

--- a/include/linker/linker-defs.h
+++ b/include/linker/linker-defs.h
@@ -115,8 +115,8 @@
    archives like KBuild did.*/
 #endif
 
-#define X(i, j)  KERNEL_OBJECT_FILE_##i (j)
-#define Y(i, j) *KERNEL_OBJECT_FILE_##i
+#define X(i, j) KERNEL_OBJECT_FILE_##i (j)
+#define Y(i, j) KERNEL_OBJECT_FILE_##i
 
 #define KERNEL_INPUT_SECTION(sect) \
     UTIL_LISTIFY(NUM_KERNEL_OBJECT_FILES, X, sect)

